core/src/autogluon/core/hpo/executors.py

```diff
@@ -0,0 +1,370 @@
+        """
+        raise NotImplementedError
+    
+    @abstractmethod
+    def register_resources(self, resources: dict):
+        """
+        Register total resources used for the experiment
+        
+        Parameters
+        ----------
+        resources
+            Total resources available for the experiment
+        """
+        raise NotImplementedError
+    
+    @abstractmethod
+    def validate_search_space(
+            self,
+            search_space: dict,
+            model_name: str,
+        ):
+        """
+        Validate the search space for the experiment
+        
+        Parameters
+        ----------
+        search_space
+            Search space provided to the experiment.
+        model_name
+            The model name of the hpo experiment is tuning. This is only used for logging purpose
+        """
+        raise NotImplementedError
+    
+    @abstractmethod
+    def execute(self, **kwargs):
+        """Execute the experiment"""
+        raise NotImplementedError
+    
+    @abstractmethod
+    def report(
+            self,
+            reporter: 'LocalReporter',
+            **kwargs
+        ):
+        """Report result of the experiment to the reporter. If no reporter needed, pass in None"""
+        raise NotImplementedError
+    
+    @abstractmethod
+    def get_hpo_results(
+            self,
+            model_name: str,
+            model_path_root: str,
+            **kwargs
+        ):
+        """
+        Retrieve hpo results
+        
+        Parameters
+        ----------
+        model_name
+            The model name of the hpo experiment is tuning.
+        model_path_root
+            Root path of the model
+        """
+        raise NotImplementedError
+
+
+class RayHpoExecutor(HpoExecutor):
+    """Implementation of HpoExecutor Interface, where ray tune is used as the backend"""
+    
+    custom_to_ray_preset_map = {
+        'auto': {'scheduler': 'FIFO', 'searcher': 'bayes'},
+        'local_random': {'scheduler': 'FIFO', 'searcher': 'random'},
+        'random': {'scheduler': 'FIFO', 'searcher': 'random'},
+    }
+    custom_to_ray_scheduler_preset_map = {
+        'local': 'FIFO',
+    }
+    custom_to_ray_searcher_preset_map = {
+        'local_random': 'random',
+        'random': 'random',
+        'auto': 'random',
+    }
+    
+    def __init__(self):
+        self.resources = None
+        self.hyperparameter_tune_kwargs = None
+        self.analysis = None
+    
+    @property
+    def executor_type(self):
+        return RAY_BACKEND
+    
+    def initialize(self, hyperparameter_tune_kwargs, default_num_trials=None, time_limit=None):
+        self.time_limit = time_limit
+        if isinstance(hyperparameter_tune_kwargs, dict):
+            hyperparameter_tune_kwargs = hyperparameter_tune_kwargs.copy()
+        if isinstance(hyperparameter_tune_kwargs, str):
+            hyperparameter_tune_kwargs = self.custom_to_ray_preset_map[hyperparameter_tune_kwargs]
```

Comment:
Missing copy()?

Reply:
This isn't a string. You are modifying the dict returned from elf.custom_to_ray_preset_map[hyperparameter_tune_kwargs] later in this method.

Reply:
You may also want to add the type hint so this is easier to understand: 

```
hyperparameter_tune_kwargs: dict = self.custom_to_ray_preset_map[hyperparameter_tune_kwargs].copy()
```

Reply:
Oh, I think you are correct. Thanks!